File: server/src/castle-wall/allowlist/rule-identity.ts
/**
 * Stable rule identity and manifest filename contract.
 *
 * MANIFEST-RULEID-PATH-01: the producer (manifest-publisher.ts) writes
 * encoded_v1 filenames (rid1_<base64url(id)>.json). Readers accept both
 * encoded_v1 and legacy_safe for backward compatibility with already-persisted
 * manifests.
 */

import { bytesToString, fromBase64urlStrict, stringToBytes, toBase64url } from "../../core/encoding.js";

declare const ruleIdBrand: unique symbol;
declare const ruleFilenameBrand: unique symbol;

export type RuleId = string & { readonly [ruleIdBrand]: "RuleId" };
export type RuleFilename = string & { readonly [ruleFilenameBrand]: "RuleFilename" };

export const RULE_ID_PATTERN = /^[A-Za-z0-9][A-Za-z0-9._:-]{0,119}$/;
export const RULE_ID_MAX_LENGTH = 120;
export const ENCODED_RULE_FILENAME_PREFIX = "rid1_";
export const RULE_FILENAME_SUFFIX = ".json";

export type RuleFilenameRelation = "encoded_v1" | "legacy_safe";

export interface ManifestRuleIdentityEntry {
  rule_id: unknown;
  file: unknown;
}

export type RuleIdentityResult<T> =
  | { ok: true; value: T }
  | { ok: false; error: string };

export function parseRuleId(value: unknown): RuleIdentityResult<RuleId> {
  if (typeof value !== "string") return { ok: false, error: "rule id must be a string" };
  if (!RULE_ID_PATTERN.test(value)) {
    return {
      ok: false,
      error: `rule id must match ${RULE_ID_PATTERN.source} and be at most ${RULE_ID_MAX_LENGTH} ASCII characters`,
    };
  }
  return { ok: true, value: value as RuleId };
}

export function validateRuleId(value: unknown): string | null {
  const result = parseRuleId(value);
  return result.ok ? null : result.error;
}

export function encodeRuleFilename(ruleId: RuleId): RuleFilename {
  return `${ENCODED_RULE_FILENAME_PREFIX}${toBase64url(stringToBytes(ruleId))}${RULE_FILENAME_SUFFIX}` as RuleFilename;
}

function parseEncodedRuleFilename(filename: string): RuleIdentityResult<RuleFilename> {
  if (!filename.startsWith(ENCODED_RULE_FILENAME_PREFIX) || !filename.endsWith(RULE_FILENAME_SUFFIX)) {
    return { ok: false, error: "not an encoded-v1 rule filename" };
  }
  const payload = filename.slice(ENCODED_RULE_FILENAME_PREFIX.length, -RULE_FILENAME_SUFFIX.length);
  if (payload.length === 0 || !/^[A-Za-z0-9_-]+$/.test(payload)) {
    return { ok: false, error: "encoded-v1 rule filename has a non-canonical base64url payload" };
  }
  try {
    const decoded = bytesToString(fromBase64urlStrict(payload));
    const ruleId = parseRuleId(decoded);
    if (!ruleId.ok || encodeRuleFilename(ruleId.value) !== filename) {
      return { ok: false, error: "encoded-v1 rule filename does not round-trip canonically" };
    }
  } catch {
    return { ok: false, error: "encoded-v1 rule filename has an invalid base64url payload" };
  }
  return { ok: true, value: filename as RuleFilename };
}

/** Classify an exact persisted `(rule_id, file)` relation. */
export function classifyManifestRuleFilename(
  ruleIdValue: unknown,
  filenameValue: unknown,
): RuleIdentityResult<RuleFilenameRelation> {
  const parsedRuleId = parseRuleId(ruleIdValue);
  if (!parsedRuleId.ok) return parsedRuleId;
  if (typeof filenameValue !== "string") return { ok: false, error: "rule filename must be a string" };

  const expectedEncoded = encodeRuleFilename(parsedRuleId.value);
  if (filenameValue === expectedEncoded) return { ok: true, value: "encoded_v1" };
  if (filenameValue === `${parsedRuleId.value}${RULE_FILENAME_SUFFIX}`) {
    return { ok: true, value: "legacy_safe" };
  }

  const encoded = parseEncodedRuleFilename(filenameValue);
  return encoded.ok
    ? { ok: false, error: "encoded-v1 rule filename does not match rule id" }
    : { ok: false, error: "rule filename is neither encoded-v1 nor the exact legacy-safe relation" };
}

/**
 * Read-only phase-one manifest scan. It intentionally consumes no referenced
 * filename and reports every malformed relation or full-set collision.
 */
export function preflightManifestRuleEntries(entriesValue: unknown): string[] {
  const issues: string[] = [];
  if (!Array.isArray(entriesValue)) return ["manifest rules must be an array"];
  const ruleIds = new Set<string>();
  const filenames = new Set<string>();
  for (const [index, entryValue] of entriesValue.entries()) {
    if (entryValue === null || typeof entryValue !== "object") {
      issues.push(`manifest rule ${index}: entry must be an object`);
      continue;
    }
    const entry = entryValue as ManifestRuleIdentityEntry;
    if (typeof entry.rule_id === "string") {
      if (ruleIds.has(entry.rule_id)) issues.push(`manifest rule ${index}: duplicate rule id`);
      ruleIds.add(entry.rule_id);
    }
    if (typeof entry.file === "string") {
      if (filenames.has(entry.file)) issues.push(`manifest rule ${index}: duplicate rule filename`);
      filenames.add(entry.file);
    }
    const relation = classifyManifestRuleFilename(entry.rule_id, entry.file);
    if (!relation.ok) {
      issues.push(`manifest rule ${index}: ${relation.error}`);
    }
  }
  return issues;
}

File: castle-wall-daemon/src/manifest/rule_identity.rs
//! Cross-language rule ID and manifest filename contract.

use base64::Engine;
use std::collections::HashSet;

use super::verify::ManifestRuleEntry;

pub const RULE_ID_MAX_LENGTH: usize = 120;
pub const ENCODED_RULE_FILENAME_PREFIX: &str = "rid1_";
pub const RULE_FILENAME_SUFFIX: &str = ".json";

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RuleFilenameRelation {
    EncodedV1,
    LegacySafe,
}

pub fn validate_rule_id(value: &str) -> Result<(), String> {
    let bytes = value.as_bytes();
    if bytes.is_empty() || bytes.len() > RULE_ID_MAX_LENGTH {
        return Err(format!(
            "rule id must be 1..={} ASCII characters",
            RULE_ID_MAX_LENGTH
        ));
    }
    if !bytes[0].is_ascii_alphanumeric() {
        return Err("rule id must begin with an ASCII alphanumeric character".to_string());
    }
    if !bytes
        .iter()
        .all(|byte| byte.is_ascii_alphanumeric() || matches!(*byte, b'.' | b'_' | b':' | b'-'))
    {
        return Err("rule id contains a character outside the ASCII contract".to_string());
    }
    Ok(())
}

pub fn encode_rule_filename(rule_id: &str) -> Result<String, String> {
    validate_rule_id(rule_id)?;
    Ok(format!(
        "{}{}{}",
        ENCODED_RULE_FILENAME_PREFIX,
        base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(rule_id.as_bytes()),
        RULE_FILENAME_SUFFIX
    ))
}

pub fn classify_manifest_rule_filename(
    rule_id: &str,
    filename: &str,
) -> Result<RuleFilenameRelation, String> {
    validate_rule_id(rule_id)?;
    if filename == encode_rule_filename(rule_id)? {
        return Ok(RuleFilenameRelation::EncodedV1);
    }
    if filename == format!("{}{}", rule_id, RULE_FILENAME_SUFFIX) {
        return Ok(RuleFilenameRelation::LegacySafe);
    }
    if filename.starts_with(ENCODED_RULE_FILENAME_PREFIX)
        && filename.ends_with(RULE_FILENAME_SUFFIX)
    {
        let payload = &filename
            [ENCODED_RULE_FILENAME_PREFIX.len()..filename.len() - RULE_FILENAME_SUFFIX.len()];
        if payload.is_empty()
            || !payload
                .bytes()
                .all(|b| b.is_ascii_alphanumeric() || b == b'_' || b == b'-')
        {
            return Err(
                "encoded-v1 rule filename has a non-canonical base64url payload".to_string(),
            );
        }
        let decoded = base64::engine::general_purpose::URL_SAFE_NO_PAD
            .decode(payload)
            .map_err(|_| "encoded-v1 rule filename has an invalid base64url payload".to_string())?;
        let decoded_id = std::str::from_utf8(&decoded)
            .map_err(|_| "encoded-v1 rule filename does not decode as UTF-8".to_string())?;
        validate_rule_id(decoded_id)?;
        if encode_rule_filename(decoded_id)? != filename || decoded_id != rule_id {
            return Err("encoded-v1 rule filename does not match rule id canonically".to_string());
        }
        return Err("encoded-v1 rule filename does not match rule id".to_string());
    }
    Err("rule filename is neither encoded-v1 nor the exact legacy-safe relation".to_string())
}

/// Pure phase-one scan: validate every persisted relation before any filename
/// becomes filesystem authority.
pub fn preflight_manifest_rule_entries(entries: &[ManifestRuleEntry]) -> Vec<String> {
    let mut issues = Vec::new();
    let mut ids = HashSet::new();
    let mut filenames = HashSet::new();
    for (index, entry) in entries.iter().enumerate() {
        if !ids.insert(&entry.rule_id) {
            issues.push(format!("manifest rule {}: duplicate rule id", index));
        }
        if !filenames.insert(&entry.file) {
            issues.push(format!("manifest rule {}: duplicate rule filename", index));
        }
        if let Err(reason) = classify_manifest_rule_filename(&entry.rule_id, &entry.file) {
            issues.push(format!("manifest rule {}: {}", index, reason));
        }
    }
    issues
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde::Deserialize;

    #[derive(Deserialize)]
    struct Vector {
        id: String,
        encoded_v1: String,
        legacy_safe: String,
    }
    #[derive(Deserialize)]
    struct Fixture {
        valid: Vec<Vector>,
        invalid_ids: Vec<String>,
        invalid_filenames: Vec<String>,
    }

    #[test]
    fn shared_contract_vectors_match_exactly() {
        let fixture: Fixture =
            serde_json::from_str(include_str!("../../test-vectors/rule-id-filename-v1.json"))
                .unwrap();
        for vector in fixture.valid {
            assert_eq!(encode_rule_filename(&vector.id).unwrap(), vector.encoded_v1);
            assert_eq!(
                classify_manifest_rule_filename(&vector.id, &vector.legacy_safe),
                Ok(RuleFilenameRelation::LegacySafe)
            );
        }
        for invalid in fixture.invalid_ids {
            assert!(validate_rule_id(&invalid).is_err());
        }
        for invalid in fixture.invalid_filenames {
            assert!(classify_manifest_rule_filename("a", &invalid).is_err());
        }
    }

    #[test]
    fn encodes_exact_url_safe_no_pad_filename() {
        assert_eq!(
            encode_rule_filename("curated:alpha_1.2-3").unwrap(),
            "rid1_Y3VyYXRlZDphbHBoYV8xLjItMw.json"
        );
    }

    #[test]
    fn rejects_invalid_ids_and_off_contract_relations() {
        for id in ["", ".leading", "has space", "a/b", "café", &"a".repeat(121)] {
            assert!(validate_rule_id(id).is_err(), "{id:?}");
        }
        assert!(classify_manifest_rule_filename("safe", "other.json").is_err());
        assert!(classify_manifest_rule_filename("safe", "rid1_c2FmZQ=.json").is_err());
    }
}
